source/common/network/io_socket_handle_impl.cc

```diff
@@ -635,8 +635,11 @@ Address::InstanceConstSharedPtr IoSocketHandleImpl::peerAddress() {
 
 void IoSocketHandleImpl::initializeFileEvent(Event::Dispatcher& dispatcher, Event::FileReadyCb cb,
                                              Event::FileTriggerType trigger, uint32_t events) {
-  ASSERT(file_event_ == nullptr, "Attempting to initialize two `file_event_` for the same "
-                                 "file descriptor. This is not allowed.");
+  // Reset the file_event_ for the case of initialize two `file_event_` for
+  // the same file descriptor. This is not allowed.
+  if (file_event_ != nullptr) {
+    file_event_.reset();
+  }
```

Comment:
Another fix will be to reset the event in the destructor of the listener filter, but that would require every listener filter to take care of it.

Also, this fix doesn't prevent initiate two instances of FileEventImpl directly, it is how the bug https://github.com/istio/istio/issues/18229 was trigger initially, but I searched the code, we don't have that case anymore. Since we create FileEventImpl only through the dispatcher. So maybe we can move the constructor of FileEventImpl as protected, and only dispatcher can instantiate it.

Reply:
I have concerns about file_event_ referring to a Event::FileReadyCb cb that points to a deleted listener filter.  Given that, I think it is important that we find a way to reset the file event when the listen filter is destroyed, rather than hack it this way.  Who owns the listener IoSocketHandleImpl?

I think that the 2 listener filters that seem to be calling initializeFileEvent are tls_inspector and http_inspector.  Is the issue here that those filters need events during some parts of the connection lifetime, but eventually need to remove themselves from accepting events directly?

It would be useful to know more about when this ASSERT fails, and possibly change this from ASSERT to RELEASE_ASSERT.  Do the added tests cause this ASSERT to be hit?

Reply:
One option that could be considered for cases like this one is to add a RAII object that can be used for filters like this one to hold the file_event registration in cases where the filter does own a the file event registration.

Also, I think that falling back to the default filter chain in cases where tls inspector times out may not be the right thing.  If the client is not sending a handshake within a reasonable time, the proxy should just close the connection.

Reply:
If this happens only on continue_on_listener_filters_timeout_, would resetting the file_event in ActiveTcpSocket::onTimeout() solve the issue?

Reply:
No, because ActiveTcpSocket doesn't know who owns the relevant connection socket.  The filters need to properly manage the file events they own.

Reply:
Agreed! That would be ideal.

Reply:
> I have concerns about file_event_ referring to a Event::FileReadyCb cb that points to a deleted listener filter. Given that, I think it is important that we find a way to reset the file event when the listen filter is destroyed, rather than hack it this way. Who owns the listener IoSocketHandleImpl?

Indeed, FileReadyCb is pointed to a deleted listener filter, I missed that. IoSocketHandleImpl is owned by AcceptedSocketImpl which is owned by ActiveTcpSocket, then passes the ownership to the ActiveTcpConnection.

> I think that the 2 listener filters that seem to be calling initializeFileEvent are tls_inspector and http_inspector. Is the issue here that those filters need events during some parts of the connection lifetime, but eventually need to remove themselves from accepting events directly?

not sure I understand your question. Those filters need to ready some data from the socket, like TLS inspect need to read the tls hello message. Those added events will be removed after the new FileEventImpl is created. https://github.com/envoyproxy/envoy/blob/b164962eb367e054df7fd563acce59663b16dd86/source/common/network/io_socket_handle_impl.cc#L640

And it will trigger the same issue https://github.com/envoyproxy/envoy/pull/8922 fixed.

> 
> It would be useful to know more about when this ASSERT fails, and possibly change this from ASSERT to RELEASE_ASSERT. Do the added tests cause this ASSERT to be hit?

I added an integration test for this case https://github.com/envoyproxy/envoy/pull/16952/files#diff-11652aeb508dd4b3ac24783eb9a5d38d7e33e42791cbce382b0d5b269797e9f1R146
If you removed the fix from this PR, that integration test will fail at the ASSERTION.

Reply:
I looked at the code of TLS inspector and I think I understand the issue better.  I have concerns about the TLS inspector filter effectively removing itself from the filter chain on timeout.  This behavior allows for trivial bypass of the TLS inspector and HTTP inspector functionality by waiting for a few seconds before sending the rest of the client hello.  Are these inspectors purely debug hooks or are they meant to provide specific functionality or security properties?

Thanks for confirming that your new tests repo the ASSERT failure.  I suggest we try to fix this issue by calling resetFileEvent in TLS inspector and HTTP inspector destructors if the filters haven't invoked their respective done callbacks.

Reply:
> I looked at the code of TLS inspector and I think I understand the issue better. I have concerns about the TLS inspector filter effectively removing itself from the filter chain on timeout. This behavior allows for trivial bypass of the TLS inspector and HTTP inspector functionality by waiting for a few seconds before sending the rest of the client hello. Are these inspectors purely debug hooks or are they meant to provide specific functionality or security properties?

Those inspectors are meant to provider specific functionality, for example, tls inspect get the SNI for later the filter chain matching.
> 
> Thanks for confirming that your new tests repo the ASSERT failure. I suggest we try to fix this issue by calling resetFileEvent in TLS inspector and HTTP inspector destructors if the filters haven't invoked their respective done callbacks.

got it, thanks for the review!

Reply:
> > I looked at the code of TLS inspector and I think I understand the issue better. I have concerns about the TLS inspector filter effectively removing itself from the filter chain on timeout. This behavior allows for trivial bypass of the TLS inspector and HTTP inspector functionality by waiting for a few seconds before sending the rest of the client hello. Are these inspectors purely debug hooks or are they meant to provide specific functionality or security properties?
> 
> Those inspectors are meant to provider specific functionality, for example, tls inspect get the SNI for later the filter chain matching.

This seems problematic since the SNI extraction logic of the filter is trivially bypass-able by waiting for a timeout condition before sening the SNI.  Is there potential for that to cause any security concerns?

> > Thanks for confirming that your new tests repo the ASSERT failure. I suggest we try to fix this issue by calling resetFileEvent in TLS inspector and HTTP inspector destructors if the filters haven't invoked their respective done callbacks.
> 
> got it, thanks for the review!

Reply:
> > Those inspectors are meant to provider specific functionality, for example, tls inspect get the SNI for later the filter chain matching.
> 
> This seems problematic since the SNI extraction logic of the filter is trivially bypass-able by waiting for a timeout condition before sening the SNI. Is there potential for that to cause any security concerns?
> 

I searched that, the Istio is using that https://github.com/istio/istio/blob/f38465f8f5c1edd944ed7776c1253ffa2d9768b2/pilot/pkg/networking/core/v1alpha3/listener_builder.go#L213

the `protocolDetectionTimeout`'s doc probably explains the reason https://istio.io/latest/docs/reference/config/istio.mesh.v1alpha1/

so it should be for a listener accepts both tls and server-first protocol connection.

Reply:
I do appreciate the desire to handle all kinds of things transparently, but adding significant delays for all connections to server-talks-first protocols seems broken.